### DISCUSSION TITLE
Noob got some Questions where i have to input my API and get the Output on my Bot

### DISCUSSION THREAD

**Discussion by SomebodyToLove1337** [microsoft/hack-together-teams#88] [Q&A] [answered]
Hello,

i just started to Code 3 Month Ago.

I´m a little bit confused with the "Getting Started" Template.
I just want to code a bot who send a random developer joke on teams after you send the command "devJoke".
Right now i found where i can change the default command and can change the text on the adaptiv card. I create a "ApiRequest.cs" File for my API Request on the root folder but i have no idea where i should call my method to get the API request and put in in a variable to output in on teams...

Maybe someone with more experiance can me explain where i have to call the method and save it to a variable to output it.
I think the easiest to help me is to send you my repository: 
https://github.com/SomebodyToLove1337/HackTheJokesOut

This is my ApiRequest.cs:
`
using System;
using System.Net.Http;
using System.Threading.Tasks;

	public class ApiRequest
	{
		private static readonly HttpClient client = new HttpClient();

		public async Task<string> GetResponseAsync(string url)
		{
			try
			{
				HttpResponseMessage response = await client.GetAsync(url);
				response.EnsureSuccessStatusCode();
				string responseBody = await response.Content.ReadAsStringAsync();
				return responseBody;
			}
			catch (HttpRequestException e)
			{
				Console.WriteLine("\nException Caught!");
				Console.WriteLine("Message :{0} ", e.Message);
				return null;
			}
		}
	}
`

And this part of the file "HelloWorldCommandHandler.cs" should be for the Output:
`
            var cardContent = new AdaptiveCardTemplate(cardTemplate).Expand

            (

                new HelloWorldModel

                {

                    Title = "You got a new DevJoke:",

                    Body = "Bla bla bla Haha",

                }

`
But i dont know where i can call my API and save the Output in a variable to output it on my Body :(

I hope you can understand what i want ;)

Sorry for the complete noob question... But i want to do it right and learn it for the future.

Thanks for any help and please explain it to me if possible dont post only some code that work :)

Greetings
Somebody

**Comment by aycabas** [comment, marked as answer]
Hi @SomebodyToLove1337, so great to have you with us! Seems like you've already done the hard part, you can call your API in the `HelloWorldCommandHandler.cs`, that's the place we return our values. If you replace **Title** and **Body** values in your model with your api response variables, you should be able to see the jokes in your adaptive card. Let me know if it works or if you need further assistance.

**Comment by SomebodyToLove1337** [comment]
Hi @aycabas,

could it be that easy? Yes it could... Thank you so mutch, i completly overthink it.... Thanks for the Tipp. Finally i got it to work!!!!

Debugging never was funnier :)
![image](https://github.com/microsoft/hack-together-teams/assets/68236595/de51cf6e-b40a-4f87-ad84-abb3de0007fe)

So my Hack is ready to submit.
But i wanna maybe add more commands for different jokes. Did you know how i can make it? 
I know that i can add more commands in my manifest file, but what i have to copy in my HelloWorldCommandHandler.cs to get different responses for the commands?

Thank you so mutch Ayca, you saved my day <3

Greetings
Benedikt

**Reply by aycabas** [reply]
So glad to hear it's working, and funny developer joke! 😆

To add other jokes, you can simply create another CommandHandler.cs which will be the copy of HelloWorldCommandHandler.cs. You'll only need to update the RegExpTrigger in your new command handler and call your new API to get a different response. 

Let me know if it works! More dev jokes, the better 🥳

**Comment by SomebodyToLove1337** [comment]
Unfortunately it does not work.
I copied the file "HelloWorldCommandHandler.cs" and changed it.
But unfortunately nothing happens when I enter my new command "dadJoke". I have set a breakpoint in the new CommandHandler.cs, but the program does not go into the code at all.

Here is still the code of my new file "DadJokeCommandHandler.cs":

```
using AdaptiveCards.Templating;
using HackTheJokesOut.Models;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Schema;
using Microsoft.TeamsFx.Conversation;
using Newtonsoft.Json;

namespace HackTheJokesOut.Commands
{
	// Get Jokes from API



	/// <summary>
	/// The <see cref="DadJokeCommandHandler"/> registers a pattern with the <see cref="ITeamsCommandHandler"/> and 
	/// responds with an Adaptive Card if the user types the <see cref="TriggerPatterns"/>.
	/// </summary>
	public class DadJokeCommandHandler : ITeamsCommandHandler
	{
		private readonly ILogger<DadJokeCommandHandler> _logger;
		private readonly string _adaptiveCardFilePath = Path.Combine(".", "Resources", "HelloWorldCard.json");

		public IEnumerable<ITriggerPattern> TriggerPatterns => new List<ITriggerPattern>
		{
            // Used to trigger the command handler if the command text contains 'helloWorld'
            new RegExpTrigger("dadJoke")
		};

		public DadJokeCommandHandler(ILogger<DadJokeCommandHandler> logger)
		{
			_logger = logger;
		}



		public async Task<ICommandResponse> HandleCommandAsync(ITurnContext turnContext, CommandMessage message, CancellationToken cancellationToken = default)
		{
			_logger?.LogInformation($"Bot received message: {message.Text}");

			// Read adaptive card template
			var cardTemplate = await File.ReadAllTextAsync(_adaptiveCardFilePath, cancellationToken);

			// API Call
			ApiRequest apiRequest = new ApiRequest();
			string response = await apiRequest.GetResponseApi("https://icanhazdadjoke.com");
			List<Joke> jokes = JsonConvert.DeserializeObject<List<Joke>>(response);
			string jokeQuestion = jokes[0].Question;
			string jokePunchline = jokes[0].Punchline;

			// Render adaptive card content
			var cardContent = new AdaptiveCardTemplate(cardTemplate).Expand
			(
				new HelloWorldModel
				{
					Title = jokeQuestion,
					Body = jokePunchline
				}
			);

			// Build attachment
			var activity = MessageFactory.Attachment
			(
				new Attachment
				{
					ContentType = "application/vnd.microsoft.card.adaptive",
					Content = JsonConvert.DeserializeObject(cardContent),
				}
			);

			// send response
			return new ActivityCommandResponse(activity);
		}


	}


}
```
Manifest File:
```
{
    "$schema": "https://developer.microsoft.com/en-us/json-schemas/teams/v1.14/MicrosoftTeams.schema.json",
    "manifestVersion": "1.14",
    "version": "1.0.0",
    "id": "{{state.fx-resource-appstudio.teamsAppId}}",
    "packageName": "com.microsoft.teams.extension",
    "developer": {
        "name": "Teams App, Inc.",
        "websiteUrl": "https://www.example.com",
        "privacyUrl": "https://www.example.com/termofuse",
        "termsOfUseUrl": "https://www.example.com/privacy"
    },
    "icons": {
        "color": "{{config.manifest.icons.color}}",
        "outline": "{{config.manifest.icons.outline}}"
    },
    "name": {
        "short": "{{config.manifest.appName.short}}",
        "full": "{{config.manifest.appName.full}}"
    },
    "description": {
        "short": "{{config.manifest.description.short}}",
        "full": "{{config.manifest.description.full}}"
    },
    "accentColor": "#FFFFFF",
    "bots": [
        {
            "botId": "{{state.fx-resource-bot.botId}}",
            "scopes": [
                "personal",
                "team",
                "groupchat"
            ],
            "supportsFiles": false,
            "isNotificationOnly": false,
            "commandLists": [
                {
                    "scopes": [
                        "personal",
                        "team",
                        "groupchat"
                    ],
                    "commands": [
                      {
                        "title": "devJoke",
                        "description": "Let´s bring a bit of fun in your dev life"
                      },
                      {
                        "title": "dadJoke",
                        "description": "Let´s bring a bit of fun in your dev life"
                      }
                    ]
                }
            ]
        }
    ],
    "composeExtensions": [],
    "configurableTabs": [],
    "staticTabs": [],
    "permissions": [
        "identity",
        "messageTeamMembers"
    ],
    "validDomains": [
        "{{state.fx-resource-bot.domain}}"
    ]
}
```

Well it's not too bad, I'm submitting the program now with just the Developer Jokes. But I will keep trying regardless of the event.

**Reply by waldekmastykarz** [reply]
Have you updated your app in Teams so that it reflects the latest changes?

**Comment by SomebodyToLove1337** [comment]
Actually I have updated everything, also I have updated all the dependencies again.
All other changes I make in the dev jokes are also applied, but I just can't get the second command to run. 

My command has also been added to the suggestions in teams:

![image](https://github.com/microsoft/hack-together-teams/assets/68236595/f89f36db-3931-4f18-996a-ff3910225f1f)

Actually I have updated everything, also I have updated all the dependencies again.
All other changes I make in the dev jokes are also applied, but I just can't get the second command to run. 

My command has also been added to the suggestions in teams:

But it's not that bad, I'll find the bug eventually :)
